File: tauri/src/lib/services/accountsApi.ts
import { serverDefaults } from "$lib/data/serverDefaults.svelte";
import { transfers } from "$lib/data/transfers.svelte";
import { authHeadersOf, backend, baseUrlOf, type Profile } from "./backend.svelte";
import { http, type HttpClient } from "./http";

export type ProviderAuthKind = "none" | "bearer" | "api_key" | "basic" | "header";

export interface ProviderAuth {
  kind: ProviderAuthKind;
  token: string;
  user: string;
  password: string;
  headerName: string;
  headerValue: string;
}

export const emptyAuth = (): ProviderAuth => ({
  kind: "none",
  token: "",
  user: "",
  password: "",
  headerName: "",
  headerValue: "",
});

const authWire = (auth: ProviderAuth): Wire => ({
  kind: auth.kind,
  token: auth.token.trim(),
  user: auth.user.trim(),
  password: auth.password,
  header_name: auth.headerName.trim(),
  header_value: auth.headerValue.trim(),
});

export interface AccountProvider {
  baseUrl: string;
  model: string;
  contextScope: string;
}

export interface ProviderSettings extends AccountProvider {
  auth: ProviderAuth;
}

export interface Account {
  id: string;
  label: string;
  loggedIn: boolean;
  primary: boolean;
  provider: AccountProvider | null;
}

export interface ProviderProbe {
  baseUrl: string;
  models: string[];
  found: boolean;
}

export interface ProviderPreset {
  id: string;
  label: string;
  baseUrl: string;
}

export interface AccountsSnapshot {
  accounts: Account[];
  default: string;
  providerUrl: string;
  presets: ProviderPreset[];
  scopes: string[];
}

type Wire = Record<string, any>;

const parse = (data: Wire): Account => ({
  id: data.id ?? "",
  label: data.label ?? "",
  loggedIn: data.logged_in === true,
  primary: data.primary === true,
  provider: data.provider
    ? {
        baseUrl: data.provider.base_url ?? "",
        model: data.provider.model ?? "",
        contextScope: data.provider.context_scope ?? "",
      }
    : null,
});

export const exportUrl = (id: string, profile: Profile = backend.active) =>
  `${baseUrlOf(profile)}/accounts/${encodeURIComponent(id)}/export`;

export const importUrl = (label: string, profile: Profile = backend.active) =>
  `${baseUrlOf(profile)}/accounts/import?label=${encodeURIComponent(label)}`;

export const createAccountsApi = (client: HttpClient) => ({
  async list(): Promise<AccountsSnapshot | null> {
    const data = await client.get<Wire>("/accounts");
    if (!data) return null;
    return {
      accounts: (data.accounts ?? []).map(parse),
      default: data.default ?? "",
      providerUrl: data.provider_url ?? "",
      presets: (data.provider_presets ?? []).map((preset: Wire) => ({
        id: preset.id ?? "",
        label: preset.label ?? "",
        baseUrl: preset.base_url ?? "",
      })),
      scopes: data.provider_scopes ?? [],
    };
  },

  async create(label: string): Promise<Account | null> {
    const data = await client.post<Wire>("/accounts", { label });
    if (data) serverDefaults.bump();
    return data && parse(data);
  },

  async detectProvider(baseUrl = "", auth = emptyAuth()): Promise<ProviderProbe | null> {
    const data = await client.post<Wire>("/accounts/provider/probe", {
      base_url: baseUrl,
      auth: authWire(auth),
    });
    if (!data) return null;
    return { baseUrl: data.base_url ?? "", models: data.models ?? [], found: data.found === true };
  },

  async provider(id: string): Promise<ProviderSettings | null> {
    const data = await client.get<Wire>(`/accounts/${id}/provider`);
    if (!data) return null;
    const auth = data.auth ?? {};
    return {
      baseUrl: data.base_url ?? "",
      model: data.model ?? "",
      contextScope: data.context_scope ?? "",
      auth: {
        kind: auth.kind ?? "none",
        token: auth.token ?? "",
        user: auth.user ?? "",
        password: auth.password ?? "",
        headerName: auth.header_name ?? "",
        headerValue: auth.header_value ?? "",
      },
    };
  },

  async updateProvider(
    id: string,
    baseUrl: string,
    auth: ProviderAuth,
    contextScope: string,
  ): Promise<boolean> {
    const body = { base_url: baseUrl, auth: authWire(auth), context_scope: contextScope };
    const done = (await client.put(`/accounts/${id}/provider`, body)) !== null;
    if (done) serverDefaults.bump();
    return done;
  },

  async createProvider(
    label: string,
    baseUrl: string,
    auth: ProviderAuth,
    contextScope: string,
  ): Promise<Account | null> {
    const data = await client.post<Wire>("/accounts/provider", {
      label,
      base_url: baseUrl,
      auth: authWire(auth),
      context_scope: contextScope,
    });
    if (data) serverDefaults.bump();
    return data && parse(data);
  },

  async rename(id: string, label: string): Promise<boolean> {
    const done = (await client.put(`/accounts/${id}`, { label })) !== null;
    if (done) serverDefaults.bump();
    return done;
  },

  async remove(id: string): Promise<boolean> {
    const done = (await client.delete(`/accounts/${id}`)) !== null;
    if (done) serverDefaults.bump();
    return done;
  },

  async startLogin(id: string): Promise<string | null> {
    const data = await client.post<Wire>(`/accounts/${id}/login`);
    return data?.url ?? null;
  },

  async submitCode(id: string, code: string): Promise<boolean> {
    const done = (await client.post(`/accounts/${id}/login/code`, { code })) !== null;
    if (done) serverDefaults.bump();
    return done;
  },

  async cancelLogin(id: string): Promise<void> {
    await client.delete(`/accounts/${id}/login`);
  },

  exportUrl,

  async importBundle(file: File, label: string): Promise<boolean> {
    return transfers.task("upload", file.name, async (_onProgress, signal) => {
      try {
        const response = await fetch(importUrl(label), {
          method: "POST",
          headers: authHeadersOf(backend.active),
          body: file,
          signal,
        });
        if (response.ok) serverDefaults.bump();
        return response.ok;
      } catch {
        return false;
      }
    });
  },
});

export const accountsApi = createAccountsApi(http);

File: client/app/src/commonMain/kotlin/com/jahirtrap/cconnect/data/remote/AccountsApi.kt
package com.jahirtrap.cconnect.data.remote

import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put

object AccountsApi {

    data class Provider(val baseUrl: String, val model: String)

    data class Account(
        val id: String,
        val label: String,
        val loggedIn: Boolean,
        val primary: Boolean,
        val provider: Provider? = null,
    )

    data class Snapshot(val accounts: List<Account>, val default: String, val providerUrl: String = "")

    data class ProviderProbe(val baseUrl: String, val models: List<String>, val found: Boolean)

    data class ProviderSettings(val baseUrl: String, val model: String, val auth: ProviderAuth)

    data class ProviderAuth(
        val kind: String = "none",
        val token: String = "",
        val user: String = "",
        val password: String = "",
        val headerName: String = "",
        val headerValue: String = "",
    )

    private fun authWire(auth: ProviderAuth) = buildJsonObject {
        put("kind", auth.kind)
        put("token", auth.token.trim())
        put("user", auth.user.trim())
        put("password", auth.password)
        put("header_name", auth.headerName.trim())
        put("header_value", auth.headerValue.trim())
    }

    suspend fun list(): Snapshot? {
        val data = Http.get("/accounts")?.jsonObject ?: return null
        return Snapshot(
            accounts = data["accounts"]?.jsonArray?.map { parse(it.jsonObject) }.orEmpty(),
            default = data["default"]?.jsonPrimitive?.contentOrNull.orEmpty(),
            providerUrl = data["provider_url"]?.jsonPrimitive?.contentOrNull.orEmpty(),
        )
    }

    suspend fun create(label: String): Account? =
        Http.post("/accounts", buildJsonObject { put("label", label) })?.jsonObject?.let(::parse)

    suspend fun detectProvider(baseUrl: String = "", auth: ProviderAuth = ProviderAuth()): ProviderProbe? {
        val body = buildJsonObject {
            put("base_url", baseUrl)
            put("auth", authWire(auth))
        }
        val data = Http.post("/accounts/provider/probe", body)?.jsonObject ?: return null
        return ProviderProbe(
            baseUrl = data["base_url"]?.jsonPrimitive?.contentOrNull.orEmpty(),
            models = data["models"]?.jsonArray?.mapNotNull { it.jsonPrimitive.contentOrNull }.orEmpty(),
            found = data["found"]?.jsonPrimitive?.booleanOrNull ?: false,
        )
    }

    suspend fun provider(id: String): ProviderSettings? {
        val data = Http.get("/accounts/$id/provider")?.jsonObject ?: return null
        val auth = data["auth"]?.jsonObject
        fun field(name: String) = auth?.get(name)?.jsonPrimitive?.contentOrNull.orEmpty()
        return ProviderSettings(
            baseUrl = data["base_url"]?.jsonPrimitive?.contentOrNull.orEmpty(),
            model = data["model"]?.jsonPrimitive?.contentOrNull.orEmpty(),
            auth = ProviderAuth(
                kind = field("kind").ifBlank { "none" },
                token = field("token"),
                user = field("user"),
                password = field("password"),
                headerName = field("header_name"),
                headerValue = field("header_value"),
            ),
        )
    }

    suspend fun updateProvider(id: String, baseUrl: String, auth: ProviderAuth): Boolean =
        Http.put("/accounts/$id/provider", buildJsonObject {
            put("base_url", baseUrl)
            put("auth", authWire(auth))
        }) != null

    suspend fun createProvider(label: String, baseUrl: String, auth: ProviderAuth): Account? =
        Http.post("/accounts/provider", buildJsonObject {
            put("label", label)
            put("base_url", baseUrl)
            put("auth", authWire(auth))
        })?.jsonObject?.let(::parse)

    suspend fun rename(id: String, label: String): Boolean =
        Http.put("/accounts/$id", buildJsonObject { put("label", label) }) != null

    suspend fun delete(id: String): Boolean = Http.delete("/accounts/$id") != null


    suspend fun startLogin(id: String): String? =
        Http.post("/accounts/$id/login")?.jsonObject?.get("url")?.jsonPrimitive?.contentOrNull

    suspend fun submitCode(id: String, code: String): Boolean =
        Http.post("/accounts/$id/login/code", buildJsonObject { put("code", code) }) != null

    suspend fun cancelLogin(id: String) {
        Http.delete("/accounts/$id/login")
    }

    fun exportUrl(id: String): String = "${Backend.baseUrl}/accounts/${UrlCodec.encode(id)}/export"

    fun importUrl(label: String): String =
        "${Backend.baseUrl}/accounts/import?label=${UrlCodec.encode(label)}"

    private fun parse(o: JsonObject) = Account(
        id = o["id"]?.jsonPrimitive?.contentOrNull.orEmpty(),
        label = o["label"]?.jsonPrimitive?.contentOrNull.orEmpty(),
        loggedIn = o["logged_in"]?.jsonPrimitive?.booleanOrNull ?: false,
        primary = o["primary"]?.jsonPrimitive?.booleanOrNull ?: false,
        provider = (o["provider"] as? JsonObject)?.let {
            Provider(
                baseUrl = it["base_url"]?.jsonPrimitive?.contentOrNull.orEmpty(),
                model = it["model"]?.jsonPrimitive?.contentOrNull.orEmpty(),
            )
        },
    )
}
